tensorflow_addons/layers/multihead_attention.py

```diff
@@ -0,0 +1,128 @@
+        self,
+        head_size,
+        num_heads,
+        output_size=None,
+        dropout_rate=None,
+        use_projection_bias=True,
+        **kwargs
+    ):
+        self.head_size = head_size
+        self.num_heads = num_heads
+        self.output_size = output_size
+        self.dropout_rate = dropout_rate
+        self.use_projection_bias = use_projection_bias
+
+        super().__init__(**kwargs)
+
+    def build(self, input_shape):
+
+        num_query_features = input_shape[0][-1]
+        num_key_features = input_shape[1][-1]
+        num_value_features = input_shape[2][-1]
+        output_size = self.output_size if self.output_size else num_value_features
+
+        self.query_kernel = self.add_weight(
+            name="query_kernel",
+            shape=[self.num_heads, num_query_features, self.head_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+        self.key_kernel = self.add_weight(
+            name="key_kernel",
+            shape=[self.num_heads, num_key_features, self.head_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+        self.value_kernel = self.add_weight(
+            name="value_kernel",
+            shape=[self.num_heads, num_value_features, self.head_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+        self.projection_kernel = self.add_weight(
+            name="projection_kernel",
+            shape=[self.num_heads, self.size_per_head, output_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+
+        if self.use_projection_bias:
+            self.projection_bias = self.add_weight(
+                name="projection_bias",
+                shape=[output_size],
+                # initializer=self.bias_initializer,
+                # regularizer=self.bias_regularizer,
+                # constraint=self.bias_constraint,
+            )
+        else:
+            self.projection_bias = None
+
+        self.dropout = tf.keras.layers.Dropout(self.dropout_rate)
+
+        super().build(input_shape)
+
+    def call(self, inputs):
+
+        # Nomenclature for einsum
+        # ------------------------
+        # N = query elements
+        # M = key/value elements
+        # H = heads
+        # I = input features
+        # O = output features
+
+        query = inputs[0]
+        key = inputs[1]
+        value = inputs[2]
+        attention_mask = inputs[3] if len(inputs) == 4 else None
+
+        # Linear transformations
+        query = tf.einsum("...NI , HIO -> ...NHO", query, self.query_kernel)
```

Comment:
Question: do you use einsum for code simplicity or because it is more efficient than matmul in this case (or both)?

Reply:
Its mostly like this right now because I based on the `tensorflow/models` implementation which used `EinsumDense` which did it like this. More concretely, I am not sure `matmul` whould preserve the `H` dimension so you would have to do some juggling which `reshape`, which `einsum` it simpler. Don't know about performance though.

Reply:
Ok, thanks for the info!

Reply:
The performance difference isn't that much. `einsum` is much better for this use case